renderdoc/driver/vulkan/vk_memory.cpp

```diff
@@ -316,6 +316,16 @@ MemoryAllocation WrappedVulkan::AllocateMemoryForResource(bool buffer, VkMemoryR
 
     // do the actual allocation
     VkResult vkr = ObjDisp(d)->AllocateMemory(Unwrap(d), &info, NULL, &chunk.mem);
+
+    if(vkr == VK_ERROR_OUT_OF_HOST_MEMORY)
+    {
+      RDCDEBUG("Allocation failed. Creating new allocation of 0x%llx bytes", ret.size);
+      info.allocationSize = ret.size;
+      chunk.size = info.allocationSize;
+      vkr = ObjDisp(d)->AllocateMemory(Unwrap(d), &info, NULL, &chunk.mem);
+      allocSize = 0;
+    }
```

Comment:
This is a very inappropriate fix. If memory is exhausted then retrying an allocation with a potentially smaller size is not addressing the problem. At best it's delaying it by a small amount with a hack, but realistically out of memory errors in any case are not currently feasible to gracefully handle. Whether by aborting the capture or anything else.

Without much more significant changes, if we run out of memory during capture the only way to address it is by running on a device with more device memory available or reducing the memory footprint of the application itself.

Reply:
For me, it fixed the problem and I was able to make many more captures without getting this error again. I got this error once when it wanted to allocate 256 MB memory but it actually needed 7 MB. (The `allocSize` is getting bigger and bigger till it's 256 MB.) I think we shouldn't use so much memory.

Reply:
A local hack can work around a problem that you're having but that doesn't make it appropriate to make as a change for everyone.

> I got this error once when it wanted to allocate 256 MB memory but it actually needed 7 MB. (The allocSize is getting bigger and bigger till it's 256 MB.) I think we shouldn't use so much memory.

This sounds like you have the wrong idea of how the code works. The code is doing pooled allocations to try and have as few separate allocations as possible, since vulkan has a limit on the number of individual allocations and applications must pool their use. When some more memory is required and all the pools are used then a new pool is allocated, in an exponential fashion up to 256 MB. This gives a tradeoff between initially small allocations if only a small amount of memory is used, then increasing to reduce the number of individual allocations as more memory is required.

So the 7 MB allocation in your case means a new 256 MB allocation is made, but then the next 249 MB are 'free' because that pool has already been created. We don't know for a particular application where it will stop so we can't allocate 'just enough' in advance. If the system is memory constrained enough and the application memory heavy enough that we skirt close to the limit, we can't know if reducing that 256 MB allocation will even help. All this change does is delay the inevitable.